### PR TITLE
Jetty 12 : Move `FileID` from `jetty-deploy` to `jetty-util`

### PR DESCRIPTION
**Pull request by joakime**
+ Move static methods from other places to `FileID` or `URIUtil`
  - `MultiReleaseJarFile`
  - `Resource`
  - `MetaInfConfiguration`
+ Improve testing of `URIUtil` and `FileID`

### CODE REVIEW

File: jetty-core/jetty-deploy/src/test/java/org/eclipse/jetty/deploy/MockAppProvider.java [hunk 1]
```diff
@@ -60,9 +60,9 @@ public ContextHandler createContextHandler(App app) throws Exception
 
         String name = app.getPath().toString();
         name = name.substring(name.lastIndexOf("-")  + 1);
-        File war = new File(webappsDir, name);
+        Path war = new File(webappsDir, name).toPath();
```
Comment (lorban) [inline jetty-core/jetty-deploy/src/test/java/org/eclipse/jetty/deploy/MockAppProvider.java]: Why don't you change the `File webappsDir` field instead?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+{
+    /**
+     * Does the provided path have a directory segment with
+     * the configured name.
+     *
+     * @param path the path to search
+     * @param directoryName the directory name to look for (case insensitive lookup)
+     * @return true if the directory name exists in path, false if otherwise
+     */
+    public static boolean containsDirectory(Path path, String directoryName)
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java]: This implementation is broken when `path` is `/`.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+    /**
+     * Retrieve the basename of a path. This is the name of the
+     * last segment of the path, with any dot suffix (e.g. ".war") removed
+     *
+     * @param path The string path
+     * @return The last segment of the path without any dot suffix
+     */
+    public static String getBasename(Path path)
+    {
+        String basename = path.getFileName().toString();
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java]: You should test for `path.getFileName()` returning `null`.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+     */
+    public static String getExtension(Path path)
+    {
+        if (path == null)
+            return null; // no path
+
+        if (!Files.isRegularFile(path))
+            return null; // not a file
+
+        return getExtension(path.getFileName().toString());
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:64]: You should test for `path.getFileName()` returning `null`.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+     * Predicate to skip {@code META-INF/versions/*} tree from walk/stream results.
+     *
+     * <p>
+     * This only works with a zipfs based FileSystem
+     * </p>
+     *
+     * @param path the path to test
+     * @return true if not in {@code META-INF/versions/*} tree
+     */
+    public static boolean skipMetaInfVersions(Path path)
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java]: Rename to `isNotMetaInfVersions`?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+     * Predicate to skip {@code module-info.class} files.
+     *
+     * <p>
+     * This is a simple test against the last path segment using {@link Path#getFileName()}
+     * </p>
+     *
+     * @param path the path to test
+     * @return true if not a {@code module-info.class} file
+     */
+    public static boolean skipModuleInfoClass(Path path)
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java]: rename to `isNotModuleInfoClass`?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java [hunk 7]
```diff
@@ -1647,25 +1655,7 @@ public static String addQueries(String query1, String query2)
...
-        // Get SSP (retaining encoded form)
-        String s = uri.getRawSchemeSpecificPart();
-        int bangSlash = s.indexOf("!/");
-        if (bangSlash >= 0)
-            s = s.substring(0, bangSlash);
-        return URI.create(s);
-    }
-
-    public static URI fixBadJavaIoFileUrl(URI uri)
+    public static URI correctBadFileURI(URI uri)
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java]: Rename to `correctFileURI`?

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java [hunk 8]
```diff
@@ -852,6 +719,7 @@ public Resource resolve(String subUriPath) throws IOException
         return newResource(resolvedUri);
     }
 
+    // TODO: move to URIUtil
     private static URI uriResolve(URI uri, String subUriPath) throws IOException
```
Comment (lorban) [inline jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: Move to `URIUtil` maybe? :-)

File: jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationParser.java [hunk 9]
```diff
@@ -917,7 +917,7 @@ public void resetParsedClasses()
      *
      * @param name the class file name
      * @return whether the class file name is valid
-     * TODO: does multiple things, we should be able to move this FileID in future PR
+     * TODO: does multiple things, we should be able to eliminate this in future PR by consolidating the ZipFS and Directory scanning behaviors into a single Files.walk() stream
```
Comment (lorban) [inline jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationParser.java:920]: Do you already have the shape of that future PR in your head?

**Review by lorban — CHANGES_REQUESTED** [review]
There are a couple small bugs with `FileID` helpers' handling of `/` when `Path.getFilename()` is called, otherwise this looks good.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java (hunk 2 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java]: This case is protected by the `path.getNameCount()` which returns 0, causing the for loop statements to not run.
I'll add a testcase for this anyway.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java (hunk 3 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java]: Adding testcase

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java (hunk 4 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:64]: Adding testcase

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java (hunk 8 above)
Comment (joakime) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java]: I'm going to get rid of this, as it doesn't make much sense anymore.

File: jetty-core/jetty-deploy/src/test/java/org/eclipse/jetty/deploy/MockAppProvider.java (hunk 1 above)
Comment (joakime) [reply jetty-core/jetty-deploy/src/test/java/org/eclipse/jetty/deploy/MockAppProvider.java]: Done.

File: jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationParser.java (hunk 9 above)
Comment (joakime) [reply jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationParser.java:920]: Yes.

I want to consolidate the WAR/JAR file scanning/walking with the Directory scanning/walking.
AFAICT, the only difference between the two is that the Directory walking technique also does a "is hidden file" check, which we have a static predicate for in `FileID`, and that same check in WAR walk is harmless.   That means the AnnotationParser gets even simpler.

File: jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java (hunk 2 above)
Comment (lorban) [reply jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java]: Wouldn't `containsDirectory(Path.get("/"), "tmp")` return false even if `/tmp` exists?

**Review by lorban — CHANGES_REQUESTED** [review]
LGTM, besides one last tiny bug.

**Review by lorban — APPROVED** [review]

File: jetty-ee10/jetty-ee10-tests/jetty-ee10-test-quickstart/src/test/resources/jetty-logging.properties [hunk 10]
```diff
@@ -1,3 +1,5 @@
 # Jetty Logging using jetty-slf4j-impl
 org.eclipse.jetty.LEVEL=INFO
+org.eclipse.jetty.xml.LEVEL=DEBUG
+org.eclipse.jetty.quickstart.LEVEL=DEBUG
```
Comment (joakime) [inline jetty-ee10/jetty-ee10-tests/jetty-ee10-test-quickstart/src/test/resources/jetty-logging.properties]: This file was restored/rollback'd in later commit fdbeb2af7f407975b1143df08d2a5e0fbbb29c93

**Review by lorban — APPROVED** [review]

**Comment by joakime** [comment]
Merged into `jetty-12.0.x`

**Pull request closed without merging**